File: desktop-app/src/store/notifications.ts
import { create } from "zustand";
import { invoke } from "@tauri-apps/api/core";
import { listen } from "@tauri-apps/api/event";

export interface NotificationAction {
  title: string;
  actionId: string;
}

export interface AppNotification {
  id: string;
  packageName: string;
  appName: string;
  title: string;
  text: string;
  subText?: string | null;
  category?: string | null;
  postedAt: number;
  actions: NotificationAction[];
  iconBase64?: string | null;
  // Android's FLAG_ONGOING_EVENT/FLAG_NO_CLEAR (e.g. a car-key connection status) — the OS won't let
  // it be dismissed remotely, so the UI shouldn't offer to.
  ongoing: boolean;
}

interface NotificationsState {
  notifications: AppNotification[];
  loaded: boolean;
  errorMessage: string | null;
}

export const useNotificationsStore = create<NotificationsState>(() => ({
  notifications: [],
  loaded: false,
  errorMessage: null,
}));

function showTransientError(message: string) {
  useNotificationsStore.setState({ errorMessage: message });
  setTimeout(() => useNotificationsStore.setState({ errorMessage: null }), 4000);
}

/** Every notifications.* command below goes through this instead of calling `invoke` directly —
 *  without it, a rejected command (most commonly "no active connection", e.g. clicking Sync with
 *  nothing paired) was an unhandled promise rejection with zero user-facing feedback: the button
 *  looked clickable but visibly did nothing. Same fix as `invokeFiles` in store/files.ts. */
async function invokeNotifications<T>(command: string, args?: Record<string, unknown>): Promise<T | undefined> {
  try {
    return await invoke<T>(command, args);
  } catch (e) {
    showTransientError(typeof e === "string" ? e : "Something went wrong — check the connection");
    return undefined;
  }
}

export async function dismissNotification(id: string) {
  // Optimistic — matches the backend's own "remove immediately, then tell the phone" order.
  useNotificationsStore.setState((s) => ({
    notifications: s.notifications.filter((n) => n.id !== id),
  }));
  await invokeNotifications("dismiss_notification", { id });
}

/** Clears every dismissible notification at once — same per-id `notification.dismiss` the backend
 *  sends for a single dismiss, just for all of them. `ongoing` notifications (a car-key connection
 *  status, etc.) are left alone: Android refuses to let a NotificationListenerService cancel those,
 *  so optimistically clearing them here would just have them reappear on the next sync. Optimistic
 *  for the ones it does clear, same ordering as the single-dismiss case. */
export async function dismissAllNotifications() {
  const dismissible = useNotificationsStore.getState().notifications.filter((n) => !n.ongoing);
  if (dismissible.length === 0) return;
  useNotificationsStore.setState((s) => ({ notifications: s.notifications.filter((n) => n.ongoing) }));
  await invokeNotifications("dismiss_all_notifications");
}

/** Asks the phone for a fresh full list of whatever's actually showing right now — recovers from
 *  a posted/removed event missed while disconnected, which the live push-only
 *  `notification-posted`/`notification-removed` listeners below have no way to catch up on by
 *  themselves. The response replaces the list outright (see the "notifications-synced" listener),
 *  same full-snapshot-replace shape as Notes' own `refreshNotes`. */
export async function refreshNotifications() {
  await invokeNotifications("refresh_notifications");
}

let initialized = false;

export function initNotificationListeners() {
  if (initialized) return;
  initialized = true;

  invoke<AppNotification[]>("list_notifications").then((notifications) => {
    useNotificationsStore.setState({ notifications, loaded: true });
  });

  listen<AppNotification>("notification-posted", (event) => {
    useNotificationsStore.setState((s) => ({
      notifications: [event.payload, ...s.notifications.filter((n) => n.id !== event.payload.id)],
    }));
  });

  listen<{ id: string }>("notification-removed", (event) => {
    useNotificationsStore.setState((s) => ({
      notifications: s.notifications.filter((n) => n.id !== event.payload.id),
    }));
  });

  // Full-snapshot replace — response to refreshNotifications(), not an incremental update like
  // the two listeners above.
  listen<AppNotification[]>("notifications-synced", (event) => {
    useNotificationsStore.setState({ notifications: event.payload });
  });
}

File: desktop-app/src-tauri/src/commands/notifications.rs
use std::sync::Arc;

use crate::net::server::{send_to_active, AppState};
use crate::protocol::envelope::{EmptyPayload, NotificationPostedPayload, NotificationRemovedPayload};

#[tauri::command]
pub async fn list_notifications(
    state: tauri::State<'_, Arc<AppState>>,
) -> Result<Vec<NotificationPostedPayload>, String> {
    Ok(state.notifications.lock().await.clone())
}

/// Asks the phone for a fresh full list of whatever's currently showing — see
/// `dispatch::notifications::sync` for why this exists (recovering from a posted/removed event
/// missed while disconnected, which the live `notification.posted`/`notification.removed` stream
/// has no way to catch up on by itself).
#[tauri::command]
pub async fn refresh_notifications(state: tauri::State<'_, Arc<AppState>>) -> Result<(), String> {
    send_to_active(&state, "notifications.refresh", &EmptyPayload {})
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn dismiss_notification(
    state: tauri::State<'_, Arc<AppState>>,
    id: String,
) -> Result<(), String> {
    state.notifications.lock().await.retain(|n| n.id != id);
    // Doesn't actually clear the native Notification Center banner today — see notify.rs's module doc.
    crate::notify::remove(&state.app_handle, &id);
    send_to_active(&state, "notification.dismiss", &NotificationRemovedPayload { id })
        .await
        .map_err(|e| e.to_string())
}

/// No bulk `notification.dismissAll` on the wire — clears every notification's phone-side copy
/// with the same per-id `notification.dismiss` message `dismiss_notification` sends, just for all
/// of them, while `notify::remove_all` clears every native banner in one call rather than one
/// `removeDeliveredNotificationsWithIdentifiers` round trip per id.
///
/// Skips `ongoing` notifications (Android's FLAG_ONGOING_EVENT/FLAG_NO_CLEAR, e.g. a car-key
/// connection status) entirely: the OS refuses to let a NotificationListenerService cancel those,
/// so sending the dismiss would just silently no-op and the next sync would bring them right back —
/// leave them in `state.notifications` and their native banner alone instead of flashing them away.
#[tauri::command]
pub async fn dismiss_all_notifications(state: tauri::State<'_, Arc<AppState>>) -> Result<(), String> {
    let dismissible_ids: Vec<String> = {
        let mut notifications = state.notifications.lock().await;
        let (dismissible, remaining): (Vec<_>, Vec<_>) =
            notifications.drain(..).partition(|n| !n.ongoing);
        *notifications = remaining;
        dismissible.into_iter().map(|n| n.id).collect()
    };
    for id in &dismissible_ids {
        crate::notify::remove(&state.app_handle, id);
    }
    let mut failed = 0usize;
    for id in &dismissible_ids {
        if send_to_active(&state, "notification.dismiss", &NotificationRemovedPayload { id: id.clone() })
            .await
            .is_err()
        {
            failed += 1;
        }
    }
    tracing::info!(
        "dismiss_all_notifications: {} sent, {} failed to send",
        dismissible_ids.len() - failed,
        failed
    );
    Ok(())
}
